Fixes #17084 - Add webpack on welcome and unauthorized pages

bastilian opened the pull request:
On pages where the welcome action in the app controller was triggered
the webpack server was not added due to welcome rendering before

bastilian committed:
Fixes #17084 - Add webpack on welcome and unauthorized pages

Adds webpack server to CSP headers via prepend_before_action,
to ensures it is run before welcome and authorized filters.

ares commented:
@theforeman-bot ok to test

ares commented:
@bastilian please change the commit message so it follows the pattern, e.g. "Fixes #17084 - add webpack on welcome packages" and `git push -f`. I tested the patch and it fixes the issue for me. If tests passes it's good to merge after the commit message change.

bastilian commented:
Commit message fixed and pushed.

domcleal commented on app/controllers/application_controller.rb:
```diff
@@ -18,8 +18,8 @@ class ApplicationController < ActionController::Base
   before_action :session_expiry, :update_activity_time, :unless => proc {|c| !SETTINGS[:login] || c.remote_user_provided? || c.api_request? }
   before_action :set_taxonomy, :require_mail, :check_empty_taxonomy
   before_action :authorize
```
The same applies to `authorize`, the CSP headers must be in place for the 403 page.

domcleal reviewed (changes requested):
Also the commit message states "welcome packages" - should it say "welcome pages"?

bastilian commented:
@domcleal packages is now corrected to pages. and it now uses `prepend_before_action` and includes a test to verify it is also set on unauthorised page requests.

domcleal commented on test/integration/middleware_test.rb:
```diff
@@ -22,4 +22,28 @@ class MiddlewareIntegrationTest < ActionDispatch::IntegrationTest
+    begin
+      Rails.configuration.webpack.dev_server.enabled = true
+      Webpack::Rails::Manifest.stubs(:asset_paths).returns([])
+      Environment.stubs(:first).returns(nil)
```
Both new tests are missing any verification that these stubs have triggered the filters you expect. Probably at a minimum they should be expectations, but some kind of verification of the page content may also help.

domcleal commented on test/integration/middleware_test.rb:
```diff
@@ -22,4 +22,28 @@ class MiddlewareIntegrationTest < ActionDispatch::IntegrationTest
+    end
+  end
+
+  test "webpack dev server adds the dev server to Content-Security-Policy on unauthirized page requests" do
```
spelling, "unauthorized"

domcleal commented on test/integration/middleware_test.rb:
```diff
@@ -22,4 +22,28 @@ class MiddlewareIntegrationTest < ActionDispatch::IntegrationTest
+    begin
+      Rails.configuration.webpack.dev_server.enabled = true
+      Webpack::Rails::Manifest.stubs(:asset_paths).returns([])
+      User.stubs(:current).returns(nil)
```
Prefer overriding `login_admin` (see user_test.rb), same comment as above applies - no verification that this tests the `authorize` filter behaviour.

bastilian commented:
@domcleal thanks for taking a closer look at it!

I've added assertions to verify the behaviour of what the stubs should trigger as well, override the `login_admin` for the unauthorised test and wrap the three tests in a context to clean it up.

domcleal commented on test/integration/middleware_test.rb:
```diff
@@ -12,14 +12,36 @@ class MiddlewareIntegrationTest < ActionDispatch::IntegrationTest
+
+      test "webpack dev server adds the dev server to Content-Security-Policy as well" do
+        visit '/environments'
+        assert page.status_code, '403'
```
`assert_equal '403', page.status_code` - this currently asserts that page.status_code is trueish

bastilian commented:
@domcleal I've cleaned it up now and added `logout_admin`, which simple removes the cookie, since overriding `login_admin` did not trigger an unauthorised request in this case.

domcleal commented:
Merged, thanks @bastilian.

bastilian commented:
Thanks for merging @domcleal! 🎉  :)
